### PR TITLE
[MERGE] mail: improve record creation with batch operations

### PR DESCRIPTION
**Pull request by seb-odoo**
PURPOSE

Optimize record creation in mail module by adding batch operations.

SPECIFICATIONS

Compute default followers once at creation for all created records. As the
same value is computed for each new record computing it once is sufficient.

Log default creation message in batch when having no specific subtype at
create. It allows to speedup record creation. For 1000 records, these specific
lines take
  * 2.1s before the commit
  * 1.2s after the commit

LINKS

PR #36654
Task 2070602

### CODE REVIEW

**Comment by seb-odoo** [comment]
@Xavier-Do if you want to have a look at what we discussed yesterday

File: addons/mail/models/mail_thread.py [hunk 1]
```diff
@@ -260,9 +260,10 @@ def create(self, vals_list):
 
         # subscribe uid unless asked not to
         if not self._context.get('mail_create_nosubscribe'):
+            default_followers = self.env['mail.followers']._add_default_followers(self._name, [], self.env.user.partner_id.ids, customer_ids=[])[0][0]
```
Comment (Xavier-Do) [inline addons/mail/models/mail_thread.py]: There is a 'conflicting wip' where we are trying to make a technical model out of mail_follower.
https://github.com/odoo/odoo/compare/master...odoo-dev:master-improve-mail-followers-jgi

Since a user cannot create a follower anymore, we cannot write on message_follower_ids and need to create them by hand using insert_followers. I also wanted to remove the 'check_exiting' in this flow since we are creating the record there, it looked useless even when giving res_ids.

Both idea could be merged together by calling inset_followers with threads.ids outside of the loop.

The previous branch was stuck because of some augmented query count because of cache invalidation linked to operation order. I was waiting for new orm to check that but I think that @tde-banana-odoo  is working on it now. 

exept that, LGTM

File: addons/test_mail/models/test_mail_models.py [hunk 2]
```diff
@@ -89,9 +89,7 @@ def _track_template(self, changes):
         return res
 
     def _creation_subtype(self):
-        if self.umbrella_id:
-            return self.env.ref('test_mail.st_mail_test_full_umbrella_upd')
-        return super(MailTestFull, self)._creation_subtype()
+        return self.env.ref('test_mail.st_mail_test_full_umbrella_upd')
```
Comment (tde-banana-odoo) [inline addons/test_mail/models/test_mail_models.py]: Why this change ?

File: addons/test_mail/tests/test_performance.py [hunk 3]
```diff
@@ -512,15 +512,15 @@ def test_complex_tracking_subscription_subtype(self):
...
+            })
 
         self.assertEqual(rec.message_partner_ids, self.partners | self.env.user.partner_id | self.user_portal.partner_id)
         # write tracking message
         self.assertEqual(rec.message_ids[0].subtype_id, self.env.ref('test_mail.st_mail_test_full_umbrella_upd'))
         self.assertEqual(rec.message_ids[0].notified_partner_ids, self.partners | self.user_portal.partner_id)
         # creation message
-        self.assertEqual(rec.message_ids[1].subtype_id, self.env.ref('mail.mt_note'))
-        self.assertEqual(rec.message_ids[1].notified_partner_ids, self.env['res.partner'])
+        self.assertEqual(rec.message_ids[1].subtype_id, self.env.ref('test_mail.st_mail_test_full_umbrella_upd'))
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_performance.py]: I would like 0 changes in test (only maybe query count) ...

File: addons/mail/models/mail_thread.py [hunk 4]
```diff
@@ -281,13 +281,13 @@ def create(self, vals_list):
         # automatic logging unless asked not to (mainly for various testing purpose)
         if not self._context.get('mail_create_nolog'):
             doc_name = self.env['ir.model']._get(self._name).name
-            for thread in threads:
-                subtype = thread._creation_subtype()
-                body = _('%s created') % doc_name
-                if subtype:  # if we have a sybtype, post message to notify users from _message_auto_subscribe
+            body = _('%s created') % doc_name
+            subtype = self._creation_subtype()
+            if subtype:  # if we have a sybtype, post message to notify users from _message_auto_subscribe
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py]: Typo

File: addons/mail/models/mail_thread.py (hunk 1 above)
Comment (tde-banana-odoo) [reply addons/mail/models/mail_thread.py]: Well first to merge wins. As it is only code move I think it is valid.

Comment (seb-odoo) [reply addons/mail/models/mail_thread.py]: I'm not sure what is the policy with the query count but it seems below expectations in almost all tests now, so maybe there doesn't need to be a new test for this case, but just lower all existing values?

**Comment by tde-banana-odoo** [comment]
Could you also create a task in SM project in order to ease followup ? Otherwise there are code changes coming out of the blue in addons we maintain.

Thanks,

File: addons/test_mail/models/test_mail_models.py (hunk 2 above)
Comment (seb-odoo) [reply addons/test_mail/models/test_mail_models.py]: Because `_creation_subtype` was made `@api.model` in order to extract it from the iteration and allow to call `_message_log_batch`.

The only place where it wasn't only model dependent was on this test.

File: addons/test_mail/tests/test_performance.py (hunk 3 above)
Comment (seb-odoo) [reply addons/test_mail/tests/test_performance.py]: Agree, 0 change in test is better, but see answer above for this specific case.

File: addons/test_mail/models/test_mail_models.py (hunk 2 above)
Comment (tde-banana-odoo) [reply addons/test_mail/models/test_mail_models.py]: Then probably your code is invalid, that's the whole purpose of that kind of tests.

Subtype creation could be dependent of your creation values. Anyway _message_log_batch should be called if no subtype is given back, as otherwise it has to be done per record.

File: addons/mail/models/mail_thread.py (hunk 1 above)
Comment (tde-banana-odoo) [reply addons/mail/models/mail_thread.py]: Do a first commit that update query count, then update it in your optimization commit so that we can see the changes.

File: addons/test_mail/models/test_mail_models.py (hunk 2 above)
Comment (seb-odoo) [reply addons/test_mail/models/test_mail_models.py]: It's convenient to drop that support since it is not used in practice.

To keep the support for this, would mean to pre-group the threads per subtype, which I wanted to avoid for simplicity of code.

But if you think it should absolutely be kept even though not used, it's fine by me, I just need to know.

Comment (tde-banana-odoo) [reply addons/test_mail/models/test_mail_models.py]: Well it does not seem costly to group threads by subtype. I would prefer keeping the feature and left test model untouched to ease cross version comparisons of mail...

File: addons/mail/models/mail_thread.py (hunk 1 above)
Comment (seb-odoo) [reply addons/mail/models/mail_thread.py]: I've reduced the number of queries to the minimum possible on local, I'll see if it works on runbot and readapt them higher if necessary.

Then I tested the current PR but both commit keep the same number of queries, they are just saving CPU time.

File: addons/test_mail/models/test_mail_models.py (hunk 2 above)
Comment (seb-odoo) [reply addons/test_mail/models/test_mail_models.py]: Adapted code to keep behavior, now doing threads without subtype in batch, after the loop.

File: addons/test_mail/tests/test_performance.py (hunk 3 above)
Comment (seb-odoo) [reply addons/test_mail/tests/test_performance.py]: Resolved

File: addons/mail/models/mail_thread.py (hunk 4 above)
Comment (seb-odoo) [reply addons/mail/models/mail_thread.py]: Done

**Comment by seb-odoo** [comment]
Thank you for the review. All suggestions have been applied.

Waiting on runbot result for query count.

**Comment by tde-banana-odoo** [comment]
Seems good, thanks for your work !

@robodoo r+ rebase-merge
